zio-logback-http4s/src/main/scala/com/softwaremill/correlator/CorrelationIdMiddleware.scala

```diff
@@ -0,0 +1,104 @@
+package com.softwaremill.correlator
+
+import java.{util => ju}
+
+import cats.data.{Kleisli, OptionT}
+import ch.qos.logback.classic.util.LogbackMDCAdapter
+import org.http4s.util.CaseInsensitiveString
+import org.http4s.{HttpRoutes, Request, Response}
+import org.slf4j.{Logger, LoggerFactory, MDC}
+import zio.random.Random
+import zio.{FiberRef, RIO, Runtime, Task, UIO, URIO, ZIO}
+import com.github.ghik.silencer.silent
+
+object CorrelationIdMiddleware {
+  import cats.implicits._
+  import zio.interop.catz._
+
+  private[correlator] final val MdcKey: String = "cid"
+  private[correlator] final val logger: Logger = LoggerFactory.getLogger(getClass.getName)
+
+  final val defaultHeaderName: String = "X-Correlation-ID"
+
+  final val defaultLogStartRequest: (String, Request[Task]) => Task[Unit] = (cid, req) =>
+    Task(logger.debug(s"Starting request with id: $cid, to: ${req.uri.path}"))
+
+  final val defaultIdGenerator: URIO[Random, String] =
+    ZIO.accessM[Random] { random =>
+      val randomUpperCaseChar: UIO[Char] = random.random.nextInt(91 - 65).map(r => (r + 65).toChar)
+      val segment: UIO[String]           = (1 to 3).toList.traverse(_ => randomUpperCaseChar).map(_.mkString)
+      (segment, segment, segment).mapN((a, b, c) => s"$a-$b-$c")
+    }
+
+  final val getCorrelationId: Task[Option[String]] = Task(Option(MDC.get(MdcKey)))
+
+  @silent("parameter value zioMDCAdapter in method addTo is never used")
+  def addTo[R <: Random](
+    headerName: String = defaultHeaderName,
+    logStartRequest: (String, Request[Task]) => Task[Unit] = defaultLogStartRequest,
+    idGenerator: RIO[R, String] = defaultIdGenerator
+  )(service: HttpRoutes[Task])(implicit runtime: Runtime[R], zioMDCAdapter: ZioMDCAdapter): HttpRoutes[Task] =
```

Comment:
One of the principal idea behind this design is that I want to force the user to initialize the `ZioMDCAdapter` before behind behing able to use the `CorrelationIdMiddleware`.

For now, to force him, I require a `ZioMDCAdapter` instance as an implicit parameter.

I don't know if it's the best possible way to do that.

Maybe, the ZIO module system can replace that unused implicit parameter by an additional constraint on `R`:  
```scala
def addTo[R <: Random with ZioMDCAdapter](...)
```
but I don't really understand how the ZIO module system work, for now.

Reply:
Indeed, the ZioMDCAdapter need not be an implicit, I didn't see where it's used by it's probably you could just write:

```scala
trait MDC {
  def mdc: MDC.Service
}
object MDC {
  trait Service { /* put MDC stuff here */ 
}
```

Then it would just work with the environment.